### DISCUSSION TITLE
How to share a httpx.AsyncClient() across the app?

### DISCUSSION THREAD

**Discussion by danechitoaie** [litestar-org/litestar#1549] [Q&A] [answered]
I need to use httpx.AsyncClient() and I've read in their docs it's more optimal to have a single client that you share across the app vs creating a client each time in the route handler.

How can I share this client so that is available across all routes and then I can safely close it on app shutdown?

**Comment by provinzkraut** [comment, marked as answer]
> and I've read in their docs it's more optimal to have a single client that you share across the app

I am not aware of this. However, there's nothing preventing you from doing that. You can just create an instance of the client and then import that.

If you want something that's specifically bound to the lifetime of the application, you could of course use application state, dependency injection and the lifetime hooks.

```python
from litestar import Litestar
from litestar.datastructures import State
from litestar.di import Provide
from httpx import AsyncClient


def on_startup(state: State) -> None:
  state.client = AsyncClient()


async def on_shutdown(state: State) -> None:
  await state.client.aclose()


def get_client(state: State) -> AsyncClient:
  return state.client


app = Litestar(
  on_startup=[on_startup],
  on_shutdown=[on_shutdown], 
  dependencies={"client": Provide(get_client)}
)
```

**Reply by danechitoaie** [reply]
> I am not aware of this.

https://www.python-httpx.org/advanced/#more-efficient-usage-of-network-resources

> On the other hand, a Client instance uses [HTTP connection pooling](https://en.wikipedia.org/wiki/HTTP_persistent_connection). This means that when you make several requests to the same host, the Client will reuse the underlying TCP connection, instead of recreating one for every single request.
> 
> This can bring significant performance improvements compared to using the top-level API, including:
> 
> - Reduced latency across requests (no handshaking).
> - Reduced CPU usage and round-trips.
> - Reduced network congestion.


I assume it's the same for AsyncClient

**Comment by WBShark** [comment]
This small code gives error for me:
  ```
await state.client.aclose()
            ^^^^^^^^^^^^
  AttributeError: 'Litestar' object has no attribute 'client'
```
I ran it with uvicorn as in most examples
What I do wrong?

PS It is my first discussion, so sorry for insufficient info
